### ISSUE TITLE
Support COLORTERM for TrueColor ASCII

### ISSUE THREAD

**Issue by atluft** [o2sh/onefetch#235]
After working on the swift logo I was thinking it might be useful to be able to specify true colors for a better result:

![image](https://user-images.githubusercontent.com/2379260/95237152-d5b88400-07d5-11eb-9c56-03a332646e45.png)

Terminals that don't support COLORTERM environment variable would look like this:
![image](https://user-images.githubusercontent.com/2379260/95237267-fa146080-07d5-11eb-83b5-c711e21e6788.png)

**Comment by o2sh** [comment]
This is very good idea @atluft 
For the final implementation, you should probably modify the `define_languages!` macro to have a second Vec! for TrueColor. By Doing that, you will be able to get rid of _the offset of 10_. Also, It's preferable to read env::var("COLORTERM") only once and pass it along.

**Comment by atluft** [comment]
Will do, new to Rust programming and having your advice helps. All comments are welcome!

**Comment by spenserblack** [comment]
@atluft Just curious: did you test how the `colored` crate renders TrueColor on a terminal that doesn't support it? I would assume that it would pick the closest available color if it can't render the TrueColor.

**Comment by spenserblack** [comment]
Nevermind, I checked the source code and it appears that `colored` *doesn't* check if COLORTERM is set or not.

I think the best way would probably be to have the `define_languages!` macro call *another* macro, like `define_colors!`, and also define a struct for colors, like
```rust
struct Colors {
    default_colors: Vec<Color>,
    true_colors: Option<Vec<Color>>,
}
```

`define_colors!` can take 2 patterns:
- `vec![ $( color:expr ),+ ]`, the current usage with no true colors, would expand to
  ```rust
  Colors {
      default_colors: vec![ $( $color ,)+ ],
      true_colors: None,
  }
  ```
- `$color_struct:expr`, which would take a definition of the `Colors` struct, and wouldn't need any special expansion rules.

Does that make sense? :smiley_cat:

**Comment by o2sh** [comment]
> did you test how the colored crate renders TrueColor on a terminal that doesn't support it

Just tested it on rxvt, it prints it in plain black:

![rxvt](https://user-images.githubusercontent.com/13710835/95335414-a9713600-08af-11eb-80df-43fe57d703bb.png)

**Comment by spenserblack** [comment]
Thanks, @o2sh!
Besides the above suggestion of adjusting the macro, we could also possibly map over defined colors and, if they're `TrueColor` and COLORTERM isn't set to an expected value, calculate the closest color ourselves using Euclidean distance (doesn't *have* to be Euclidean distance, but it's the easiest way to get closest color IMO).

**Comment by spenserblack** [comment]
I made mackwic/colored#90 for this. If it doesn't get merged, feel free to copy the code into this project :smiley:

**Comment by o2sh** [comment]
> I made mackwic/colored#90 for this.

I agree, that would be nice to have this behavior - fallback to closest Color when TrueColor isn't supported - natively handled by `colored`

Still, impressive stuff @spenserblack

**Comment by atluft** [comment]
I definitely am not rusty enough for this one :-)

I will noodle around with @o2sh suggestion above another day, it looks like a good direction.

In what I tried today, I had a hard time getting expr tokens to accept a comma suffix followed by a string literal for the languages like "c++".  I tried layering the parser so there were several rules for different formats, but it seemed like I could never get all the languages to match a rule or when I did I got ```error: variable 'x' is still repeating at this depth``` which I have no idea how to resolve.

End of my day here, have a good day where ever you are!

**Comment by spenserblack** [comment]
>I got `error: variable 'x' is still repeating at this depth`

That means there's an un-expanded repeating macro pattern somewhere. [Here is a minimal reproduction that hopefully clears it up](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=macro_rules!%20foo%20%7B%0A%20%20%20%20(%20%24(%20%24tokens%3Att%20)*%20)%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20%24tokens%20%2F%2F%20comment%20out%20this%0A%20%20%20%20%20%20%20%20%2F%2F%20%24(%20%24tokens%20)*%20%2F%2F%20and%20un-comment%20this%0A%20%20%20%20%7D%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20foo!%20%7B%0A%20%20%20%20%20%20%20%20println!(%22Hello%2C%20World!%22)%3B%0A%20%20%20%20%7D%0A%7D) (Rust's compiler errors for macros aren't the best 🙂 ).

>I definitely am not rusty enough for this one :-)

No worries, everyone starts as a beginner :smiley: :+1:

**Comment by atluft** [comment]
Thank you @spenserblack, that was a good example. I was thinking the other way around that there weren't enough matches, your example helps clearly see that the ```still repeating``` aspect is very important.

**Comment by atluft** [comment]
Not as clean as @o2sh posted and the new macro ```define_colors``` is definitely not an easy read. Most of the non-obvious coding was to skirt some error that I couldn't figure out. Either type not in agreement in the macro or rules not matching in the input data.

I begin to wonder, why isn't the color processing in the ascii file?
```
{255,0,255} .....
```

**Comment by o2sh** [comment]
> I begin to wonder, why isn't the color processing in the ascii file?

It makes it easier to manage colors - all the coloring is defined in one file (language.rs) - and allows for run-time customization via `-c, --ascii-colors <ascii-colors>`
